File: internal/capability/normalize_test.go
package capability

import (
	"strings"
	"testing"

	"github.com/ro-ag/ptrack/internal/model"
)

func TestNormalizeHTTPDisplaysCanonicalEffectiveScope(t *testing.T) {
	preview, err := Normalize(model.Capability{
		Name: " API ", Kind: model.CapabilityHTTP, AgentProfile: "agent-codex",
		Audit: model.CapabilityAuditPolicy{Enabled: true},
		HTTP: &model.HTTPScope{
			BaseURL:      "HTTPS://Example.COM:443/api/./v1/",
			Methods:      []string{"get", "HEAD", "get"},
			PathPrefixes: []string{"/api/v1/users", "/api/v1"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if preview.Capability.HTTP.BaseURL != "https://example.com/api/v1" {
		t.Errorf("base URL = %q", preview.Capability.HTTP.BaseURL)
	}
	if got := strings.Join(preview.Capability.HTTP.Methods, ","); got != "GET,HEAD" {
		t.Errorf("methods = %q", got)
	}
	if preview.ScopeDigest == "" || !strings.Contains(preview.EffectiveScope, "paths=/api/v1,/api/v1/users") {
		t.Errorf("preview = %+v", preview)
	}
	for _, field := range []string{"profile=agent-codex", "approval_duration_seconds=3600", "max_request_bytes=", "max_redirects=0", "audit enabled=true"} {
		if !strings.Contains(preview.EffectiveScope, field) {
			t.Errorf("effective scope missing %q: %s", field, preview.EffectiveScope)
		}
	}
}

func TestNormalizePreservesExplicitZeroRedirectLimit(t *testing.T) {
	preview, err := Normalize(model.Capability{
		Name: "no redirects", Kind: model.CapabilityHTTP, AgentProfile: "agent-codex",
		Limits: model.CapabilityLimits{MaxRedirects: 0},
		HTTP:   &model.HTTPScope{BaseURL: "https://example.com", Methods: []string{"GET"}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if preview.Capability.Limits.MaxRedirects != 0 || !strings.Contains(preview.EffectiveScope, "max_redirects=0") {
		t.Fatalf("preview = %+v", preview)
	}
}

func TestNormalizeRejectsCredentialsAndAmbiguousHTTPPaths(t *testing.T) {
	for _, rawURL := range []string{
		"https://token@example.com/api",
		"https://example.com/api/%2e%2e/admin",
		"https://example.com/api%2fadmin",
	} {
		_, err := Normalize(model.Capability{
			Name: "bad", Kind: model.CapabilityHTTP, AgentProfile: "agent-codex",
			HTTP: &model.HTTPScope{BaseURL: rawURL, Methods: []string{"GET"}},
		})
		if err == nil {
			t.Errorf("Normalize(%q) unexpectedly succeeded", rawURL)
		}
	}
}

func TestNormalizeHTTPPathUsesSegmentBoundaries(t *testing.T) {
	_, err := Normalize(model.Capability{
		Name: "bad path", Kind: model.CapabilityHTTP, AgentProfile: "agent-codex",
		HTTP: &model.HTTPScope{BaseURL: "https://example.com/api", Methods: []string{"GET"}, PathPrefixes: []string{"/apix"}},
	})
	if err == nil {
		t.Fatal("expected /apix to be outside /api")
	}
}

func TestNormalizeGitRejectsCredentialsAndRiskyRefspecs(t *testing.T) {
	base := model.Capability{
		Name: "repo", Kind: model.CapabilityGit, AgentProfile: "agent-codex",
		Git: &model.GitScope{RemoteName: "origin", RemoteURL: "https://example.com/repo.git", Operations: []string{"push"}},
	}
	credentialed := base
	credentialed.Git = &model.GitScope{RemoteName: "origin", RemoteURL: "https://token@example.com/repo.git", Operations: []string{"fetch"}}
	if _, err := Normalize(credentialed); err == nil {
		t.Fatal("credentialed Git URL accepted")
	}
	forced := base
	forced.Git = &model.GitScope{RemoteName: "origin", RemoteURL: "git@example.com:org/repo.git", Operations: []string{"push"}, Refspecs: []string{"+main:main"}}
	if _, err := Normalize(forced); err == nil {
		t.Fatal("force refspec accepted without grant")
	}
}

func TestNormalizeGitRejectsBranchForceTagAndPseudorefBypasses(t *testing.T) {
	for _, branch := range []string{"+main", "refs/tags/v1", "HEAD", "@", "FETCH_HEAD"} {
		_, err := Normalize(model.Capability{
			Name: "repo", Kind: model.CapabilityGit, AgentProfile: "agent-codex",
			Git: &model.GitScope{
				RemoteName: "origin", RemoteURL: "https://example.com/repo.git",
				Operations: []string{"push"}, Branches: []string{branch},
			},
		})
		if err == nil {
			t.Errorf("branch bypass %q was accepted", branch)
		}
	}
}

func TestNormalizeGitEffectiveScopeDisplaysEveryWriteGrant(t *testing.T) {
	preview, err := Normalize(model.Capability{
		Name: "repo", Kind: model.CapabilityGit, AgentProfile: "agent-codex",
		Git: &model.GitScope{
			RemoteName: "origin", RemoteURL: "https://example.com/repo.git",
			Operations: []string{"push"}, Branches: []string{"main"},
			Refspecs: []string{"main:main"}, AllowForcePush: true,
			AllowDeleteRefs: true, AllowTags: true,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	for _, field := range []string{`refspecs=["refs/heads/main:refs/heads/main"]`, "allow_tags=true", "allow_force_with_lease=true", "allow_delete_refs=true"} {
		if !strings.Contains(preview.EffectiveScope, field) {
			t.Errorf("effective scope missing %q: %s", field, preview.EffectiveScope)
		}
	}
}

func TestNormalizeGitRejectsUnconditionalForceAndUnsupportedRefNamespaces(t *testing.T) {
	for _, refspec := range []string{"+main:main", "refs/notes/review:refs/notes/review"} {
		_, err := Normalize(model.Capability{
			Name: "repo", Kind: model.CapabilityGit, AgentProfile: "agent-codex",
			Git: &model.GitScope{
				RemoteName: "origin", RemoteURL: "https://example.com/repo.git",
				Operations: []string{"push"}, Branches: []string{"main"},
				Refspecs: []string{refspec}, AllowForcePush: true,
			},
		})
		if err == nil {
			t.Errorf("ambiguous refspec %q was accepted", refspec)
		}
	}
}

func TestNormalizeSSHSeparatesHighRiskGrants(t *testing.T) {
	key := "ssh-ed25519 " + "QUJDREVGR0hJSktMTU5PUA=="
	preview, err := Normalize(model.Capability{
		Name: "deploy", Kind: model.CapabilitySSH, AgentProfile: "agent-codex",
		SSH: &model.SSHScope{
			Host: "EXAMPLE.com.", User: "deploy", HostKey: key,
			AllowUpload: true, UploadRoots: []string{"dist"}, UploadRemoteRoots: []string{"/srv/app"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if preview.Capability.SSH.Host != "example.com" || preview.Capability.SSH.Port != 22 {
		t.Errorf("SSH scope = %+v", preview.Capability.SSH)
	}
	if !strings.Contains(preview.EffectiveScope, `grants=["upload"]`) || strings.Contains(preview.EffectiveScope, `"interactive-shell"`) {
		t.Errorf("effective scope = %q", preview.EffectiveScope)
	}
}

func TestNormalizeSSHEffectiveScopeDisplaysExactHighRiskTargets(t *testing.T) {
	preview, err := Normalize(model.Capability{
		Name: "deploy", Kind: model.CapabilitySSH, AgentProfile: "agent-codex",
		SSH: &model.SSHScope{
			Alias: "prod", Host: "example.com", User: "deploy",
			HostKey:        "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA==",
			RemoteCommands: []string{"systemctl status app"},
			AllowUpload:    true, UploadRoots: []string{"dist"}, UploadRemoteRoots: []string{"/srv/app"},
			LocalForwardTargets: []string{"db.internal:5432"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	for _, field := range []string{`alias="prod"`, `commands=["systemctl status app"]`, `upload_local_roots=["dist"]`, `upload_remote_roots=["/srv/app"]`, `local_forward_targets=["db.internal:5432"]`} {
		if !strings.Contains(preview.EffectiveScope, field) {
			t.Errorf("effective scope missing %q: %s", field, preview.EffectiveScope)
		}
	}
}

func TestNormalizeSSHRejectsTraversalAndUnpairedTransferGrant(t *testing.T) {
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	for _, scope := range []*model.SSHScope{
		{Host: "example.com", User: "deploy", HostKey: key, AllowUpload: true, UploadRoots: []string{"../secret"}, UploadRemoteRoots: []string{"/srv/app"}},
		{Host: "example.com", User: "deploy", HostKey: key, UploadRoots: []string{"dist"}, UploadRemoteRoots: []string{"/srv/app"}},
	} {
		_, err := Normalize(model.Capability{Name: "bad", Kind: model.CapabilitySSH, AgentProfile: "agent-codex", SSH: scope})
		if err == nil {
			t.Errorf("scope %+v unexpectedly accepted", scope)
		}
	}
}

func TestNormalizeSSHRejectsUnavailableInteractiveShellGrant(t *testing.T) {
	_, err := Normalize(model.Capability{
		Name: "shell", Kind: model.CapabilitySSH, AgentProfile: "agent-codex",
		SSH: &model.SSHScope{
			Host: "example.com", User: "deploy", HostKey: "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA==",
			AllowInteractiveShell: true,
		},
	})
	if err == nil || !strings.Contains(err.Error(), "unavailable") {
		t.Fatalf("interactive-shell grant error = %v", err)
	}
}

File: crates/ptrack-capability-policy/src/normalize_test.rs
use ptrack_core::{
    Capability, CapabilityAuditPolicy, CapabilityKind, CapabilityLimits, GitScope, HttpScope,
    SshScope, Timestamp,
};

use super::*;

type InvalidDraftCase = (fn(&mut Capability), &'static str);

pub(super) fn draft(kind: CapabilityKind) -> Capability {
    Capability {
        id: 0,
        model_version: 0,
        revision: 0,
        name: "fixture".to_owned(),
        kind,
        agent_profile: "agent-codex".to_owned(),
        enabled: false,
        approval_duration_seconds: 0,
        approved_at: Timestamp::Zero,
        expires_at: Timestamp::Zero,
        scope_digest: ptrack_core::Digest32::EMPTY,
        limits: CapabilityLimits {
            timeout_seconds: 0,
            max_request_bytes: 0,
            max_response_bytes: 0,
            max_output_bytes: 0,
            max_redirects: 0,
            max_concurrent: 0,
        },
        audit: CapabilityAuditPolicy {
            enabled: false,
            retain_last: 0,
        },
        http: None,
        git: None,
        ssh: None,
        created_at: Timestamp::Zero,
        updated_at: Timestamp::Zero,
    }
}

#[test]
fn http_normalizes_defaults_lists_and_ambiguous_paths() {
    let mut value = draft(CapabilityKind::Http);
    value.name = " API ".to_owned();
    value.audit.enabled = true;
    value.http = Some(HttpScope {
        base_url: "HTTPS://Example.COM:443/api/./v1/".to_owned(),
        methods: vec!["get".to_owned(), "HEAD".to_owned(), "get".to_owned()],
        path_prefixes: vec!["/api/v1/users".to_owned(), "/api/v1".to_owned()],
    });
    let preview = normalize(&value).unwrap();
    let scope = preview.capability.http.unwrap();
    assert_eq!(scope.base_url, "https://example.com/api/v1");
    assert_eq!(scope.methods, ["GET", "HEAD"]);
    assert!(preview.effective_scope.contains("max_redirects=0"));

    for raw in [
        "https://token@example.com/api",
        "https://example.com/api/%2e%2e/admin",
        "https://example.com/api%2fadmin",
        "https://example.com/api/%00",
    ] {
        let mut candidate = draft(CapabilityKind::Http);
        candidate.http = Some(HttpScope {
            base_url: raw.to_owned(),
            methods: vec!["GET".to_owned()],
            path_prefixes: Vec::new(),
        });
        assert!(normalize(&candidate).is_err(), "accepted {raw}");
    }
}

#[test]
fn git_rewrites_bare_refspecs_and_denies_bypasses() {
    let mut value = draft(CapabilityKind::Git);
    value.git = Some(GitScope {
        remote_name: "origin".to_owned(),
        remote_url: "https://example.com:443/repo.git".to_owned(),
        operations: vec!["push".to_owned()],
        branches: vec!["main".to_owned()],
        refspecs: vec!["main:main".to_owned()],
        allow_tags: false,
        allow_force_push: true,
        allow_delete_refs: false,
    });
    let preview = normalize(&value).unwrap();
    let scope = preview.capability.git.unwrap();
    assert_eq!(scope.remote_url, "https://example.com/repo.git");
    assert_eq!(scope.refspecs, ["refs/heads/main:refs/heads/main"]);

    for branch in ["+main", "refs/tags/v1", "HEAD", "@", "FETCH_HEAD"] {
        let mut candidate = value.clone();
        candidate.git.as_mut().unwrap().branches = vec![branch.to_owned()];
        assert!(normalize(&candidate).is_err(), "accepted {branch}");
    }
}

#[test]
fn ssh_requires_independent_paired_grants_and_rejects_line_separators() {
    let mut value = draft(CapabilityKind::Ssh);
    value.ssh = Some(SshScope {
        alias: "prod".to_owned(),
        host: "EXAMPLE.com.".to_owned(),
        port: 0,
        user: "deploy".to_owned(),
        host_key: "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA==".to_owned(),
        allow_git: false,
        remote_commands: vec!["uptime".to_owned()],
        allow_upload: true,
        allow_download: false,
        upload_roots: vec!["dist".to_owned()],
        download_roots: Vec::new(),
        upload_remote_roots: vec!["/srv/app".to_owned()],
        download_remote_roots: Vec::new(),
        allow_interactive_shell: false,
        local_forward_targets: Vec::new(),
        remote_forward_targets: Vec::new(),
    });
    let preview = normalize(&value).unwrap();
    assert_eq!(preview.capability.ssh.unwrap().host, "example.com");

    value.ssh.as_mut().unwrap().remote_commands = vec!["printf  ".to_owned()];
    assert!(normalize(&value).is_err());
    value.ssh.as_mut().unwrap().remote_commands.clear();
    value.ssh.as_mut().unwrap().alias = "prod/escape".to_owned();
    assert!(normalize(&value).is_err());
}

#[test]
fn canonical_json_matches_go_html_and_line_separator_escaping() {
    let value = vec!["<>&   "];
    assert_eq!(
        super::normalize::go_json(&value).unwrap(),
        "[\"\\u003c\\u003e\\u0026 \\u2028\\u2029\"]"
    );
}

#[test]
fn draft_contract_errors_defaults_bounds_and_exclusivity() {
    let mut value = draft(CapabilityKind::Http);
    value.http = Some(HttpScope {
        base_url: "https://example.com".to_owned(),
        methods: vec!["GET".to_owned()],
        path_prefixes: Vec::new(),
    });
    let preview = normalize(&value).unwrap();
    assert_eq!(preview.capability.model_version, 1);
    assert_eq!(preview.capability.approval_duration_seconds, 3_600);
    assert_eq!(preview.capability.limits.timeout_seconds, 30);
    assert_eq!(preview.capability.limits.max_request_bytes, 1 << 20);
    assert_eq!(preview.capability.limits.max_response_bytes, 4 << 20);
    assert_eq!(preview.capability.limits.max_output_bytes, 1 << 20);
    assert_eq!(preview.capability.limits.max_redirects, 0);
    assert_eq!(preview.capability.limits.max_concurrent, 1);
    assert_eq!(preview.capability.audit.retain_last, 100);

    let cases: &[InvalidDraftCase] = &[
        (
            |candidate: &mut Capability| candidate.model_version = 2,
            "unsupported capability model version",
        ),
        (
            |candidate: &mut Capability| candidate.name.clear(),
            "capability name",
        ),
        (
            |candidate: &mut Capability| candidate.agent_profile = "-agent".to_owned(),
            "agent profile",
        ),
        (
            |candidate: &mut Capability| candidate.approval_duration_seconds = 59,
            "approval duration",
        ),
        (
            |candidate: &mut Capability| candidate.limits.timeout_seconds = 301,
            "timeout",
        ),
        (
            |candidate: &mut Capability| candidate.limits.max_request_bytes = (32 << 20) + 1,
            "maximum request",
        ),
        (
            |candidate: &mut Capability| candidate.limits.max_redirects = 11,
            "maximum redirects",
        ),
        (
            |candidate: &mut Capability| candidate.limits.max_concurrent = 9,
            "maximum concurrent",
        ),
        (
            |candidate: &mut Capability| candidate.audit.retain_last = 1_001,
            "audit retention",
        ),
    ];
    for (mutate, message) in cases {
        let mut candidate = value.clone();
        mutate(&mut candidate);
        assert!(
            normalize(&candidate)
                .unwrap_err()
                .to_string()
                .contains(message)
        );
    }
    let mut cross_kind = value.clone();
    cross_kind.git = Some(GitScope {
        remote_name: "origin".to_owned(),
        remote_url: "https://example.com/repo.git".to_owned(),
        operations: vec!["fetch".to_owned()],
        branches: Vec::new(),
        refspecs: Vec::new(),
        allow_tags: false,
        allow_force_push: false,
        allow_delete_refs: false,
    });
    assert_eq!(
        normalize(&cross_kind).unwrap_err().to_string(),
        "HTTP capability must contain only an HTTP scope"
    );
    let mut unicode_profile = value.clone();
    unicode_profile.agent_profile = "agent-١".to_owned();
    assert!(normalize(&unicode_profile).is_ok());
    unicode_profile.agent_profile = "agent-²".to_owned();
    assert!(normalize(&unicode_profile).is_err());
}

#[test]
fn strict_url_parser_rejects_whatwg_coercions_and_decodes_git_path_once() {
    for raw in [
        "example.com/api",
        "http:example.com/api",
        "https://@example.com/api",
        "https://127.1/api",
        "https://0x7f000001/api",
        "https://aé.example/api",
    ] {
        assert!(
            super::normalize::normalize_http_url(raw, false).is_err(),
            "accepted {raw}"
        );
    }
    for raw in [
        "https:example.com/repo.git",
        "https://example.com/a/%2e%2e/repo.git",
        "ssh://git@example.com/a%5Crepo.git",
        "ssh://git@example.com/a%2500repo.git",
    ] {
        assert!(
            super::normalize::normalize_git_remote(raw).is_err(),
            "accepted {raw}"
        );
    }
}
